### PR TITLE
Fixes #11930: Add a way to prevent reporting of methods

### PR DESCRIPTION
**Pull request by VinceMacBuche**
https://www.rudder-project.org/redmine/issues/11930

### CODE REVIEW

File: tree/20_cfe_basics/log_rudder.cf [hunk 1]
```diff
@@ -202,3 +224,18 @@ bundle agent clean_reporting_context {
     "report_data.rule_id"        string => "";
     "report_data.identifier"     string => "";
 }
+
+bundle agent report_data {
+  vars:
+    "should_report" string => "true";
```
Comment (peckpeck) [inline tree/20_cfe_basics/log_rudder.cf]: true and false are not classes in cfengine, use any and a class that is garanteed to be never defined

Comment (peckpeck) [reply tree/20_cfe_basics/log_rudder.cf]: or use a cmp() in the class definition in generic methods

File: tree/20_cfe_basics/log_rudder.cf [hunk 2]
```diff
@@ -107,7 +111,25 @@ bundle agent _rudder_common_reports_generic(technique_name, class_prefix, identi
                 action  => immediate_ignore_dry_run,
              ifvarclass => "(dry_run|global_dry_run).${class_prefix}_repaired";
 
-    !(dry_run|global_dry_run)::
+    !report::
```
Comment (peckpeck) [inline tree/20_cfe_basics/log_rudder.cf]: I don't understand the duplication of report calls

File: tree/20_cfe_basics/log_rudder.cf (hunk 1 above)
Comment (VinceMacBuche) [reply tree/20_cfe_basics/log_rudder.cf]: "true" and "false" can be used to defined classes like done in the other bundle

https://docs.cfengine.com/docs/3.10/reference-promise-types-classes.html#expression

Comment (peckpeck) [reply tree/20_cfe_basics/log_rudder.cf]: your first url point to nothing.
My cfengine test code says it doesn't work:

  reports:
      "Done";
    true::
      "true";
    false::
      "false";

-> Done

Comment (VinceMacBuche) [reply tree/20_cfe_basics/log_rudder.cf]: Here we are defining a variable not a class

Comment (VinceMacBuche) [reply tree/20_cfe_basics/log_rudder.cf]: You should do likewise

Comment (VinceMacBuche) [reply tree/20_cfe_basics/log_rudder.cf]: You should do this way:

vars:

  "should_report" string => "true";

classes:
  "report" expression => "${should_report}";

reports:

  report:: 
    "hey I have reporting based on ${should_report};

Comment (VinceMacBuche) [reply tree/20_cfe_basics/log_rudder.cf]: bundle agent test {

vars:

  "should_report" string => "true";

classes:
  "report" expression => "true";

reports:

  "no  reporting based on ${should_report}";
  report::
  "hey I have reporting based on ${should_report}";
}

Comment (VinceMacBuche) [reply tree/20_cfe_basics/log_rudder.cf]: here we only define the value "should_report"

Comment (VinceMacBuche) [reply tree/20_cfe_basics/log_rudder.cf]: This works since cfengine 3.9

Comment (VinceMacBuche) [reply tree/20_cfe_basics/log_rudder.cf]: All versions of Rudder (4.1 + are using cfengine 3.10)

Comment (VinceMacBuche) [reply tree/20_cfe_basics/log_rudder.cf]: So we have to chose what we want to de here

**Comment by VinceMacBuche** [comment]
PR rebased

**Comment by VinceMacBuche** [comment]
@peckpeck If ncf renaming is merged, maybe we can merge this one. Attention: there is still some methods to modify but i think we should merge this first and male another pr to fix all others

File: tree/20_cfe_basics/log_rudder.cf [hunk 3]
```diff
@@ -107,7 +111,25 @@ bundle agent _rudder_common_reports_generic(technique_name, class_prefix, identi
...
+      "repaired"
+        usebundle  => _rudder_common_report("${technique_name}", "log_repaired", "${identifier}", "${component_name}", "${component_key}", "${message_prefix} was repaired"),
+        ifvarclass => "${class_prefix}_repaired.!${class_prefix}_error";
+
+      "error"
+        usebundle  => _rudder_common_report("${technique_name}", "log_warn", "${identifier}", "${component_name}", "${component_key}", "${message_prefix} could not be repaired"),
+        ifvarclass => "${class_prefix}_error";
+
+
+    report&!(dry_run|global_dry_run)::
```
Comment (amousset) [inline tree/20_cfe_basics/log_rudder.cf]: we usually use . for "and"

File: tree/30_generic_methods/service_ensure_running.cf [hunk 4]
```diff
@@ -37,12 +37,14 @@ bundle agent service_ensure_running(service_name)
...
 
   methods:
+    "inner"     usebundle => calling_inner_method;
     "action"             usebundle => service_started("${service_name}");
     "class copy"         usebundle => _classes_copy("service_started_${canonified_service_name}", "${old_class_prefix}"),
                          ifvarclass => "service_started_${canonified_service_name}_reached";
     "new result classes" usebundle => _classes_copy("${class_prefix}_action", "${class_prefix}"),
                          ifvarclass => "${class_prefix}_action_reached";
     "report"
+    "end_inner"     usebundle => end_calling_inner_method;
```
Comment (amousset) [inline tree/30_generic_methods/service_ensure_running.cf]: the is inserted in the middle of a promise

**Comment by VinceMacBuche** [comment]
Commit modified

File: tree/30_generic_methods/service_stop.cf [hunk 5]
```diff
@@ -39,14 +39,25 @@ bundle agent service_stop(service_name)
     "class_prefix"      string => canonify(join("_", "promisers"));
     "args"                     slist => { "${service_name}" };
 
+  class:
+
+    "is_inner_method" expression => "!${report_data.should_report}";
+
   methods:
+    !is_inner_method::
```
Comment (amousset) [inline tree/30_generic_methods/service_stop.cf]: why should service_action ever make reports here?

Comment (VinceMacBuche) [reply tree/30_generic_methods/service_stop.cf]: In fact the condition is only needed on the end side we should always disable inner reporting, if it is already an inner method

**Comment by VinceMacBuche** [comment]
Commit modified

File: tree/30_generic_methods/service_stop.cf [hunk 6]
```diff
@@ -39,14 +39,24 @@ bundle agent service_stop(service_name)
     "class_prefix"      string => canonify(join("_", "promisers"));
     "args"                     slist => { "${service_name}" };
 
+  class:
+
+    "is_inner_method" expression => "!${report_data.should_report}";
```
Comment (peckpeck) [inline tree/30_generic_methods/service_stop.cf]: Shouldn't we do that on ever method ? since any method can become a inner method

Comment (VinceMacBuche) [reply tree/30_generic_methods/service_stop.cf]: Maybe, I suggest to do likewise in another PR :)

Comment (peckpeck) [reply tree/30_generic_methods/service_stop.cf]: ok, in this case, this PR is OK for me

**Comment by VinceMacBuche** [comment]
Commit modified

**Comment by VinceMacBuche** [comment]
OK, merging this PR

**Pull request merged**